mail/dovecot/files/dovecot-core-pushnotify.patch

```diff
@@ -55,619 +54,56 @@ diff -Naur ./README.md ./README.md
-++	ssize_t				rc				= 0;
-++	socklen_t			sock_len		= 0;
-++	struct sockaddr_un	sock_addr;
-++	struct msg_data_s	msg_data;
-++
-++	if (ctx->rcpt_user->mail_debug) {
-++		debug = 1;
-++		i_info( "push-notify: push notification enabled" );
-++	}
-++
-++	if ( strcasecmp(mailbox, "INBOX") != 0) {
-++		i_info( "push-notify: message saved to mailbox: %s, no notification sent", mailbox );
-++		return;
-++	}
-++
-++	notify_sock = socket( AF_UNIX, SOCK_DGRAM, 0 );
-++	if ( notify_sock < 0 ) {
-++		/* warn that connect failed but do not fail the plugin or message will not get delivered */
-++		i_warning( "push-notify: open socket: \"%s\" failed", sock_path );
-++		return;
-++	}
-++
-++	sock_addr.sun_family = AF_UNIX;
-++	strncpy( sock_addr.sun_path, "@PREFIX@/var/run/dovecot/push_notify", sizeof(sock_addr.sun_path) );
-++	sock_len = sizeof(sock_addr.sun_family) + strlen(sock_addr.sun_path) + 1;
-++	rc = connect(notify_sock, (struct sockaddr *) &sock_addr,  sock_len);
-++	if ( rc < 0 ) {
-++		/* warn that connect failed but do not fail the plugin or message will not get delivered */
-++		i_warning( "push-notify: connect() to socket: \"%s\" failed: %m", sock_path );
-++		return;
-++	}
-++
-++	memset( &msg_data, 0, sizeof( msg_data ) );
-++	msg_data.msg = 3;
-++
-++	/* set user/account id */
-++	if ( ctx->rcpt_user->username != NULL ) {
-++		strncpy( msg_data.d1, ctx->rcpt_user->username, sizeof(msg_data.d1) );
-++		if (debug)
-++			i_info( "push-notify: notify: %s", msg_data.d1 );
-++	}
-++
-++	rc = send(notify_sock, (void *)&msg_data, sizeof(struct msg_data_s), 0);
-++	if ( rc < 0 )
-++		i_warning( "push-notify: send() to socket: \"%s\" failed: %m", sock_path );
-++
-++	if (debug)
-++		i_info("push-notify: data sent: %lu", rc);
-++
-++	close(notify_sock);
-++} // push_notification
-++
-++static void push_notify_deliver(struct mail_deliver_context *ctx,
-++				const char *mailbox)
-++{
-++	push_notification(ctx, mailbox);
-++	if (next_deliver_mail != NULL)
-++		next_deliver_mail(ctx, mailbox);
-++}
-++
-++void push_notify_plugin_init (struct module *module ATTR_UNUSED)
-++{
-++	next_deliver_mail = deliver_hook;
-++	deliver_hook = push_notify_deliver;
-++}
-++
-++void push_notify_plugin_deinit (void)
-++{
-++	deliver_hook = next_deliver_mail;
-++}
-+diff -Naur ./src/plugins/push-notify/push-notify-plugin.h ./src/plugins/push-notify/push-notify-plugin.h
-+--- ./src/plugins/push-notify/push-notify-plugin.h	1969-12-31 19:00:00.000000000 -0500
-++++ ./src/plugins/push-notify/push-notify-plugin.h	2019-10-08 20:53:46.000000000 -0400
-+@@ -0,0 +1,20 @@
-++/* Copyright (c) 2008-2011 Apple, inc. */
-++
-++#ifndef __NOTIFY_PLUGIN_H__
-++#define __NOTIFY_PLUGIN_H__
-++
-++struct module;
-++void push_notify_plugin_init(struct module *module);
-++void push_notify_plugin_deinit(void);
-++
-++typedef struct msg_data_s {
-++	unsigned long msg;
-++	unsigned long pid;
-++
-++	char d1[128];
-++	char d2[512];
-++	char d3[512];
-++	char d4[512];
-++} msg_data_t;
-++
-++#endif
```

Comment:
@essandess: it looks like this was added by accident?

Reply:
This code is necessary for the `apns` variant:

https://github.com/macports/macports-ports/blob/26f5adbdc746e77befbb60579b5bc94ade1c660c/mail/dovecot/Portfile#L240-L339

It's only useful for people who go through the trouble outlines in the notes of keeping their APNS mail push notification certificates up-to-date, so should not be added unconditionally, only as part of the `apns` variant. 

It looks like the thing to do is to update this patch by updating dovecot core and computing a new diff that adds this code. See: https://github.com/essandess/core

I haven't had to update these patches at all because they've all worked with the latest dovecot updates.

Is there a reason the `apns` patches are being updated in this PR?

Reply:
I understand that there are patches needed for the `apns` variant, but your patch file is recursive. It first adds a patch for `configure.ac`, but then also creates a **_new_** file `dovecot-core-pushnotify.patch` to patch `configure.ac` in the same way. I'm referring to the super long block that I deleted from `dovecot-core-pushnotify.patch`. Everything from the deleted block was still kept as a regular patch elsewhere in that file. This block can go, right?

> It's only useful for people who go through the trouble outlined in the notes of keeping their APNS mail push notification certificates up-to-date, so should not be added unconditionally, only as part of the apns variant.

While they might only be useful for these people: do the patches hurt the others who don't want to go through the troubles outlined in the notes (and don't activate the `apns` variant)? I'm asking because it's all too easy to update the port without testing all the variant combinations, and miss the fact that the patch no longer applies.

> It looks like the thing to do is to update this patch by updating dovecot core and computing a new diff that adds this code. See: https://github.com/essandess/core

I already did that, and I would be grateful if you were willing to test this PR. (Well, I manually changed the diff rather than patching dovecot and creating a new diff, but that doesn't make any difference.)

> Is there a reason the `apns` patches are being updated in this PR?

* The patch no longer applies to the latest version (for this reason it is really necessary).
* There is a long recursive section (which is likely harmless, but super super confusing) that called for deletion.
* I changed updated line number to make the patch apply cleanly (which is again not strictly necessary, but it's nice to do).
* I removed some bogus timestamps like `1969-12-31 19:00:00.000000000 -0500` (which is not needed to be done at all, but these timestamps are not useful for anything either, and not accurate either).

Reply:
Yes, it is needed that the `apns` patch is updated. Without Mojca's fixes, dovecot 2.3.16 will not compile because the existing `apns` patch fails.

Reply:
@mojca @essandess I am now running dovecot 2.3.16 with APNS based on the way Mojca created the patch. So, I can report it works. 

I did encounter the problem again where pushnotify.pl of APNS freezes the file system of my macOS Mojave + Server. When this happens, everything that requires file system access locks up. E.g. a simple 'ls' command will just hang. In the past, I had a period I could not use push notifications because this happened. When I was slowly bringing the mail server live, it happened when I loaded the dovecot-apns launch daemon. My suspicion then was that it had to do with the kernel filesystem subsystem locking up in some way, the culprit being the socket used for APNS in combination with how my setup is configured. For instance, I tend to run things in chroot jails when possible for security reasons. My postfix, for instance, runs chrooted. pushnotify.pl also drops privileges. When I ran pushnotify.pl as root, I did not have this problem. It also might have to do with the fact that pushnotify.pl runs from launchd and starts waiting for access.

So, I needed to hard reset my server 10 minutes ago again, and after the reboot, the system is running, and APNS works, but I am somewhat nervous as in the past I had this 'file system locking up' problem too and it was triggered not immediately. The last year I have been running without trouble and my guess was that for this not to happen, I needed to make sure that the socket for APNS had the right permissions.

My working notes say:

```
BUGHUNTING push notfications

I've seen different permissions on this socket, but this works

bash-3.2# ls -l /opt/local/var/run/dovecot/push_notify
srw-rw-rw-  1 root  wheel  0 Nov 28 15:21 /opt/local/var/run/dovecot/push_notify

This works with pushnotify dropping privileges:

bash-3.2# ls -al /opt/local/etc/dovecot-apns/
total 16
drwxr-x---   5 _dovecot  _dovenull   160 Nov 29 13:48 .
drwxr-xr-x  44 root      wheel      1408 Nov 23 17:20 ..
-rw-r--r--   1 root      _dovenull     0 Nov 23 16:36 .turd_dovecot
-rw-r--r--   1 _dovecot  _dovenull  1952 Sep 13 15:23 com.apple.mail.cert.pem
-rw-r-----   1 _dovecot  _dovenull  1704 Sep 13 15:23 com.apple.mail.key.pem

```

Reply:
Okay got it, thanks.

This will require more work on my end because the commits that modify dovecot to use APNS conflict with the latest version. I'll do my best to fix these, generate a new patch file (also without the recursion issue), make sure it works, and post back.

Also, there's an alternate APNS solution that could be used if these Apple-based dovecot patches ever become to difficult to merge with the latest dovecot version: [dovecot-xaps-plugin](https://github.com/st3fan/dovecot-xaps-daemon/issues/46).

> ```
> git pull --rebase --autostash origin master
> From github.com:essandess/core
 > * branch                  master     -> FETCH_HEAD
> Auto-merging src/plugins/Makefile.am
> CONFLICT (content): Merge conflict in src/plugins/Makefile.am
> Auto-merging src/lib-lda/mail-deliver.h
> Auto-merging src/lib-lda/mail-deliver.c
> Auto-merging src/imap/imap-settings.c
> CONFLICT (content): Merge conflict in src/imap/imap-settings.c
> Auto-merging src/imap/imap-client.c
> Auto-merging src/imap/Makefile.am
> Auto-merging src/imap-login/imap-login-settings.c
> CONFLICT (content): Merge conflict in src/imap-login/imap-login-settings.c
> Auto-merging src/imap-login/imap-login-client.c
> Auto-merging src/auth/auth-settings.c
> CONFLICT (content): Merge conflict in src/auth/auth-settings.c
> Auto-merging configure.ac
> Auto-merging README.md
> error: could not apply edacd60ff7... Dovecot core Apple Push Notification Service (APNS) Plugin
> Resolve all conflicts manually, mark them as resolved with
> "git add/rm <conflicted_files>", then run "git rebase --continue".
> You can instead skip this commit: run "git rebase --skip".
> To abort and get back to the state before "git rebase", run "git rebase --abort".
> Could not apply edacd60ff7... Dovecot core Apple Push Notification Service (APNS) Plugin
> ```

Reply:
Just to make sure that we’re all on the same page with the APNS patch, please note:
* I have been maintaining this patch with a branch `apns` off a clone of dovecot core that has the working commits. See https://github.com/essandess/core/tree/apns. The MacPorts patch file for the `apns` variant is made from a diff between master and this branch.
* Up until the last version, these `apns` commits did not conflict with dovecot core updates.
* There are now several conflicts in the actual code base, possibly suggesting that the patch file mods in the current PR are inadequate. I haven’t dug into the code yet to determine of these are trivial or substantive conflicts.

Reply:
> Okay got it, thanks.
> 
> This will require more work on my end because the commits that modify dovecot to use APNS conflict with the latest version. I'll do my best to fix these, generate a new patch file (also without the recursion issue), make sure it works, and post back.

OK, but can you please at least go carefully through my notes?

The patch above **ALREADY** works. It has already been adapted to the latest release. You are just hitting the exact same conflicts that I already resolved. I can "port" them to your fork of dovecot if you want (I guess you want/need to do that for the future anyway), but there's nothing else that needs to be done about this particular PR other than testing and potentially fixing further issues that might have popped up with the latest version.

(The fastest way for you to "rebase" the changes might be to apply the patch from this PR.)

There are two types of conflicts, both trivial. One file was removed from `Makefile.am` and there was a minor change in the syntax that occurs at two locations.

> Also, there's an alternate APNS solution that could be used if these Apple-based dovecot patches ever become to difficult to merge with the latest dovecot version: [dovecot-xaps-plugin](https://github.com/st3fan/dovecot-xaps-daemon/issues/46).

I leave that up to you to decide.

One further question: is this port a good candidate for port abandonement procedure?

> I have been maintaining this patch with a branch apns off a clone of dovecot core that has the working commits. See https://github.com/essandess/core/tree/apns. The MacPorts patch file for the apns variant is made from a diff between master and this branch.

I didn't know that earlier, but I can give you a commit that fixes that.

> Up until the last version, these apns commits did not conflict with dovecot core updates.
>
> There are now several conflicts in the actual code base

Yes, precisely, this is what me and Gerben have been trying to explain to you since the beginning of the thread.

> possibly suggesting that the patch file mods in the current PR are inadequate.

The patch in this PR should work. What I cannot say is whether there were any further incompatibilities introduced in the latest version beyond what has already been patched be you before.

Btw: did you try to submit the patch upstream?

> I haven’t dug into the code yet to determine of these are trivial or substantive conflicts.

They are all straightforward. See above.

Reply:
> @mojca @essandess I am now running dovecot 2.3.16 with APNS based on the way Mojca created the patch. So, I can report it works.

That's great news, thank you.

> I did encounter the problem again where pushnotify.pl of APNS freezes the file system of my macOS Mojave + Server.

From what I understand that's unrelated to the update though, right?

Reply:
No, I am not certain it is unrelated to the update. The update may have reverted something to a situation which triggers the 'filesystem deadlock' issue on my Mojave system.

I recall that there were two approaches to APNS. I also recall that Apple was planning to change the API and retire something, which required a different approach. Getting that different solution to work was something @essandess did. So, I have been wondering if my current situation has reverted to something that can trigger the problem again. So it might be related. It is good that @essandess has a look at it to check if my current situation (with your cleaned patch) is APNS-wise like it was supposed to be with 2.3.13.

My guess is that the problem is triggered by the socket in combination with the fact that privileges are dropped and that that leads to some sort of deadlock, but I have never been able to create the exact situation that will reliably trigger it and I am loath to do so as this is a main production system which is also used for backups for instance.

Reply:
But it _is_ important for @essandess to know that my current situation with 2.3.16 and APNS does work and does provide push notifications.

Reply:
> that Apple was planning to change the API and retire something, which required a different approach.

No, Apple simply dropped support of the legacy binary protocol to its APNS API. This is independent of the two approaches, and the `apns` variant of this port supports the new API.

https://developer.apple.com/news/?id=c88acm2b

Reply:
> But it _is_ important for @essandess to know that my current situation with 2.3.16 and APNS does work and does provide push notifications.

I confirm that APNS works with this update on my mail server. The issues you refer to look like they may be on your end. I recommend comparing your configuration with what the Portfile specifies and/or chasing down logs.